From https://gitlab.alpinelinux.org/alpine/apk-tools/-/merge_requests/357
From 3726535965b8c557b60361f88f6873320b0c4299 Mon Sep 17 00:00:00 2001
From: Dimitri John Ledkov <dimitri.ledkov@chainguard.dev>
Date: Sun, 2 Nov 2025 13:35:15 +0000
Subject: [PATCH] openssl: remove deprecated API usage

With this changes apk can be compiled without any deprecated APIs.
---
 libfetch/common.c | 12 ++----------
 src/Makefile      |  4 ++--
 src/apk.c         | 23 -----------------------
 3 files changed, 4 insertions(+), 35 deletions(-)

diff --git a/libfetch/common.c b/libfetch/common.c
index af064b87..5336ec6a 100644
--- a/libfetch/common.c
+++ b/libfetch/common.c
@@ -584,15 +584,7 @@ static int fetch_ssl_setup_client_certificate(SSL_CTX *ctx, int verbose)
 int
 fetch_ssl(conn_t *conn, const struct url *URL, int verbose)
 {
-	/* Init the SSL library and context */
-	if (!SSL_library_init()){
-		fprintf(stderr, "SSL library init failed\n");
-		return (-1);
-	}
-
-	SSL_load_error_strings();
-
-	conn->ssl_meth = SSLv23_client_method();
+	conn->ssl_meth = TLS_client_method();
 	conn->ssl_ctx = SSL_CTX_new(conn->ssl_meth);
 	SSL_CTX_set_mode(conn->ssl_ctx, SSL_MODE_AUTO_RETRY);
 
@@ -620,7 +612,7 @@ fetch_ssl(conn_t *conn, const struct url *URL, int verbose)
 		return (-1);
 	}
 
-	conn->ssl_cert = SSL_get_peer_certificate(conn->ssl);
+	conn->ssl_cert = SSL_get1_peer_certificate(conn->ssl);
 	if (!conn->ssl_cert) {
 		fprintf(stderr, "No server SSL certificate\n");
 		return -1;
diff --git a/src/Makefile b/src/Makefile
index a01f5dc6..8a2ec0cb 100644
--- a/src/Makefile
+++ b/src/Makefile
@@ -73,8 +73,8 @@ LIBS_apk.so		:= $(libapk_so)
 
 CFLAGS_ALL		+= -D_ATFILE_SOURCE -Ilibfetch
 CFLAGS_apk.o		:= -DAPK_VERSION=\"$(FULL_VERSION)\"
-CFLAGS_apk-static.o	:= -DAPK_VERSION=\"$(FULL_VERSION)\" -DOPENSSL_NO_ENGINE
-CFLAGS_apk-test.o	:= -DAPK_VERSION=\"$(FULL_VERSION)\" -DOPENSSL_NO_ENGINE -DTEST_MODE
+CFLAGS_apk-static.o	:= -DAPK_VERSION=\"$(FULL_VERSION)\"
+CFLAGS_apk-test.o	:= -DAPK_VERSION=\"$(FULL_VERSION)\" -DTEST_MODE
 
 progs-$(STATIC)		+= apk.static
 apk.static-objs		:= $(filter-out apk.o,$(apk-objs)) apk-static.o
diff --git a/src/apk.c b/src/apk.c
index 3b4ce40c..164acea2 100644
--- a/src/apk.c
+++ b/src/apk.c
@@ -20,9 +20,6 @@
 #include <sys/stat.h>
 
 #include <openssl/crypto.h>
-#ifndef OPENSSL_NO_ENGINE
-#include <openssl/engine.h>
-#endif
 
 #include <fetch.h>
 
@@ -421,25 +418,6 @@ static int parse_options(int argc, char **argv, struct apk_applet *applet, void
 	return 0;
 }
 
-static void fini_openssl(void)
-{
-	EVP_cleanup();
-#ifndef OPENSSL_NO_ENGINE
-	ENGINE_cleanup();
-#endif
-	CRYPTO_cleanup_all_ex_data();
-}
-
-static void init_openssl(void)
-{
-	atexit(fini_openssl);
-	OpenSSL_add_all_algorithms();
-#ifndef OPENSSL_NO_ENGINE
-	ENGINE_load_builtin_engines();
-	ENGINE_register_all_complete();
-#endif
-}
-
 static void on_sigwinch(int s)
 {
 	apk_reset_screen_width();
@@ -531,7 +509,6 @@ int main(int argc, char **argv)
 		if (applet->update_cache) dbopts.cache_max_age = 0;
 	}
 
-	init_openssl();
 	setup_automatic_flags();
 	fetchTimeout = 60;
 	fetchRedirectMethod = fetch_redirect;
-- 
GitLab

